src/typings/formatters.ts

```diff
@@ -1,39 +1,126 @@
  *
- * @property formatString A Logback-like format string. The format string may include specifiers
- * indicating how and what kv-pair may be inserted to replace the specifier in the string. A
- * specifier uses the following syntax: `%<name>{<format-options>}`
- * - <name> is the specifier's name.
- * - <format-options> (along with the braces) are optional and indicate how to format the kv-pair
- * when inserting it. Options cannot include literal '%' or '}' characters.
- * - Literal '%' characters cannot be escaped.
- *
- * The following specifiers are currently supported:
- * - `d` - Indicates the log event's authoritative timestamp. Its options may include a
- * java.time.format.DateTimeFormatter pattern indicating how to format the timestamp. NOTE: Not
- * all patterns are supported; see `convertDateTimeFormatterPatternToDayJs` to determine what's
- * supported.
- * - `n` - Indicates a newline character. This is currently ignored by the formatter since it always
- * inserts a newline as the last character.
- * - `<key>` - Any specifier besides those above indicate a key for a kv-pair, if said kv-pair
- * exists in a given log event.
- */
-interface LogbackFormatterOptionsType {
+ * All three as may contain any character, except that colons (:), right braces (}),
+ * and backslashes (\) must be escaped with a backslash.
+ */
+interface YscopeFormatterOptionsType {
     formatString: string,
 }
 
-type FormatterOptionsType = LogbackFormatterOptionsType;
+type FormatterOptionsType = YscopeFormatterOptionsType;
 
 interface Formatter {
+
+    /**
+     * Formats the given log event.
+     *
+     * @param logEvent
+     * @return The formatted log event.
+     */
     formatLogEvent: (logEvent: LogEvent) => string
 }
 
+interface YScopeFieldFormatter {
+
+    /**
+     * Formats the given field.
+     *
+     * @param logEvent
+     * @return The formatted field.
+     */
+    formatField: (field: JsonValue) => string
+}
+
+/**
+ * Parsed field placeholder from a Yscope format string.
+ */
+type YScopeFieldPlaceholder = {
+    fieldNameKeys: string[],
+    fieldFormatter: Nullable<YScopeFieldFormatter>,
+
+    // Location of field placeholder in format string including braces.
+    range: {start: number, end: number}
+}
+
+/**
+ * Type for list of currently supported Yscope field formatters.
+ */
+type YScopeFieldFormatterMap = {
+    [key: string]: new (options: Nullable<string>) => YScopeFieldFormatter;
+};
+
+/**
+ * Rare character to replace escaped backslash in format string. Rare character is unlikely to be in
+ * user format string.  Writing regex to distinguish between a single escape character ("\") and an
+ * escaped backslash ("\\") is challenging especially when they are in series. It is simpler to just
+ * replace escaped backslashes with a rare character and add them back after parsing field
+ * placeholder with regex is finished.
+ */
```

Comment:
If you agree, the rationale behind this can be explained where it is referenced.
```suggestion
/**
 * Unicode REPLACEMENT CHARACTER `U+FFFD` to substitute escaped backslash (`\\`) in format string.
 */
```